src/System.Buffers/src/System/Buffers/Binary/Reader.cs

```diff
@@ -0,0 +1,125 @@
+        public static sbyte ReverseEndianness(sbyte value)
+        {
+            return value;
+        }
+
+        /// <summary>
+        /// Reverses a primitive value - performs an endianness swap
+        /// </summary> 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static short ReverseEndianness(short value)
+        {
+            return (short)((value & 0x00FF) << 8 | (value & 0xFF00) >> 8);
+        }
+
+        /// <summary>
+        /// Reverses a primitive value - performs an endianness swap
+        /// </summary> 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int ReverseEndianness(int value) => (int)ReverseEndianness((uint)value);
+
+        /// <summary>
+        /// Reverses a primitive value - performs an endianness swap
+        /// </summary> 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static long ReverseEndianness(long value) => (long)ReverseEndianness((ulong)value);
+
+        /// <summary>
+        /// This is a no-op and added only for consistency.
+        /// This allows the caller to read a struct of numeric primitives and reverse each field
+        /// rather than having to skip byte fields.
+        /// </summary> 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static byte ReverseEndianness(byte value)
+        {
+            return value;
+        }
+
+        /// <summary>
+        /// Reverses a primitive value - performs an endianness swap
+        /// </summary> 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ushort ReverseEndianness(ushort value)
+        {
+            return (ushort)((value & 0x00FFU) << 8 | (value & 0xFF00U) >> 8);
+        }
+
+        /// <summary>
+        /// Reverses a primitive value - performs an endianness swap
+        /// </summary> 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint ReverseEndianness(uint value)
+        {
+            value = (value << 16) | (value >> 16);
+            value = (value & 0x00FF00FF) << 8 | (value & 0xFF00FF00) >> 8;
+            return value;
+        }
+
+        /// <summary>
+        /// Reverses a primitive value - performs an endianness swap
+        /// </summary> 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ulong ReverseEndianness(ulong value)
+        {
+            value = (value << 32) | (value >> 32);
+            value = (value & 0x0000FFFF0000FFFF) << 16 | (value & 0xFFFF0000FFFF0000) >> 16;
+            value = (value & 0x00FF00FF00FF00FF) << 8 | (value & 0xFF00FF00FF00FF00) >> 8;
+            return value;
+        }
+
+        /// <summary>
+        /// Reads a structure of type T out of a read-only span of bytes.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static T ReadCurrentEndianness<T>(ReadOnlySpan<byte> buffer)
```

Comment:
Should we verify that T is a "primitive"/"blittable" and fail fast if it's not?

Reply:
Yes, it is a good idea to the approximate blittable check here - just like what we have done for Span: https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Span.cs#L87
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/Span.cs#L100

Since you are insisting on having this in separate .dll from Span, it means that there needs to be a private copy of the SpanHelpers machinery compiled into the portable System.Memory to perform this check...

Reply:
BTW, I am not insisting on System.Buffers.dll. We discussed this at the API review and I think most people felt it's safer in System.Buffers.dll (because maybe we will not in-box the types there ever). If we are sure there is no difference from OOB perspective, I don't think anybody had any stron preferences between these two dlls.

Reply:
System.Buffers is inbox in .NET Core, so it has to play by the OOB rules already. 

Actually, what happens when I am going to reference System.Buffers with these additions in my .NET Core 2.0 app? Is it going to work - what build of System.Buffers I am going to get?